include/wil/registry_helpers.h

```diff
@@ -1218,6 +1218,531 @@ namespace wil
 #endif // #if defined(WIL_ENABLE_EXCEPTIONS)
         }
 
+        // registry iterator support requires exceptions
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        namespace reg_iterator_details
+        {
+            constexpr uint32_t iterator_end_offset = 0xffffffff;
+            constexpr size_t iterator_default_buffer_size = 16;
+
+            // function overloads to allow *_enumerator objects to be constructed from all 3 types of HKEY representatives
+            inline HKEY get_hkey(HKEY h) WI_NOEXCEPT
+            {
+                return h;
+            }
+            inline HKEY get_hkey(const ::wil::unique_hkey& h) WI_NOEXCEPT
+            {
+                return h.get();
+            }
+#if defined(__WIL_WINREG_STL)
+            inline HKEY get_hkey(const ::wil::shared_hkey& h) WI_NOEXCEPT
+            {
+                return h.get();
+            }
+#endif // #if defined(__WIL_WINREG_STL)
+
+            // function overloads to allow *_iterator_data objects to hold different types to contain the string value
+            // if more string values are desired, all 5 functions must be implement for that type:
+            // clear_name, copy_name, compare_name, resize_name, address_of_name
+            inline void clear_name(::wil::unique_process_heap_string& name, size_t length)
+            {
+                if (length > 0)
+                {
+                    memset(name.get(), 0, length * sizeof(wchar_t));
+                }
+            }
+            inline ::wil::unique_process_heap_string copy_name(const ::wil::unique_process_heap_string& str, size_t capacity)
+            {
+                if (!str)
+                {
+                    return {};
+
+                }
+                return ::wil::make_process_heap_string(str.get(), capacity);
+            }
+            inline bool compare_name(const ::wil::unique_process_heap_string& name, PCWSTR comparand) WI_NOEXCEPT
+            {
+                if (!name)
+                {
+                    return false;
+                }
+                return 0 == wcscmp(name.get(), comparand);
+            }
+            inline size_t resize_name(::wil::unique_process_heap_string& name, size_t current_length, size_t new_length)
+            {
+                if (new_length > current_length)
+                {
+                    auto newString{ ::wil::make_unique_string_nothrow<::wil::unique_process_heap_string>(name.get(), new_length) };
+                    name.swap(newString);
+                    return new_length;
+                }
+
+                // continue to use the existing buffer since the requested length is less than or equals to the current length
+                clear_name(name, current_length);

```

Comment:
I'm confused. Why the call to `clear_name` here when the above (allocation code path) copies the string with a new length?